toolchain/parse/handle_declaration_scope_loop.cpp

```diff
@@ -22,75 +22,106 @@ static auto HandleUnrecognizedDecl(Context& context) -> void {
-      if (context.packaging_state() !=
-          Context::PackagingState::AfterNonPackagingDecl) {
-        if (!context.first_non_packaging_token().is_valid()) {
-          context.set_first_non_packaging_token(*context.position());
-        }
-        context.set_packaging_state(
-            Context::PackagingState::AfterNonPackagingDecl);
+  }
+
+  // Because a non-packaging keyword was encountered, packaging is complete.
+  // Misplaced packaging keywords may lead to this being re-triggered.
+  if (context.packaging_state() !=
+      Context::PackagingState::AfterNonPackagingDecl) {
+    if (!context.first_non_packaging_token().is_valid()) {
+      context.set_first_non_packaging_token(*context.position());
+    }
+    context.set_packaging_state(Context::PackagingState::AfterNonPackagingDecl);
+  }
+
+  // Remaining keywords are only valid after imports are complete, and so all
+  // result in a `set_packaging_state` call. Note, this may not always be
+  // necessary but is probably cheaper than validating.
+  switch (position_kind) {
+    case Lex::TokenKind::Abstract: {
+      if (context.PositionIs(Lex::TokenKind::Class, 1)) {
+        context.PushState(State::TypeAfterIntroducerAsClass);
+        auto abstract_token = context.Consume();
+        auto class_token = context.Consume();
+        context.AddLeafNode(NodeKind::ClassIntroducer, class_token);
+        context.AddLeafNode(NodeKind::AbstractSpecifier, abstract_token);
+        return;
       }
-      switch (position_kind) {
-        // Remaining keywords are only valid after imports are complete, and
-        // so all result in a `set_packaging_state` call. Note, this may not
-        // always be necessary but is probably cheaper than validating.
-        case Lex::TokenKind::Class: {
-          context.PushState(State::TypeIntroducerAsClass);
-          break;
-        }
-        case Lex::TokenKind::Constraint: {
-          context.PushState(State::TypeIntroducerAsNamedConstraint);
-          break;
-        }
-        case Lex::TokenKind::Fn: {
-          context.PushState(State::FunctionIntroducer);
-          break;
-        }
-        case Lex::TokenKind::Interface: {
-          context.PushState(State::TypeIntroducerAsInterface);
-          break;
-        }
-        case Lex::TokenKind::Namespace: {
-          context.PushState(State::Namespace);
-          break;
-        }
-        case Lex::TokenKind::Semi: {
-          context.AddLeafNode(NodeKind::EmptyDecl, context.Consume());
-          break;
-        }
-        case Lex::TokenKind::Var: {
-          context.PushState(State::VarAsDecl);
-          break;
-        }
-        case Lex::TokenKind::Let: {
-          context.PushState(State::Let);
-          break;
-        }
-        default: {
-          HandleUnrecognizedDecl(context);
-          break;
-        }
+      break;
+    }
+    case Lex::TokenKind::Base: {
+      if (context.PositionIs(Lex::TokenKind::Class, 1)) {
```

Comment:
Can `abstract` and `base` share their implementation? They look very similar, only differing on the specifier node kind.

Reply:
Nice, done.